ocaml/src/gblob_access.ml

```diff
@@ -0,0 +1,350 @@
+        GMemPool.free bytes;
+        Lwt.return_unit
+      )
+
+
+  method _push_blob_data fnr len slices _f =
+    let fn = self # _get_file_path fnr in
+
+    Prelude.with_timing_lwt
+      (fun () ->IOExecFile.file_open service_handle fn [Unix.O_RDONLY] )
+    >>= fun (took, handle) ->
+    AsdStatistics.new_delta _statistics _FILE_OPEN took;
+
+    let corrections =
+      List.map
+        (fun (off,len) ->
+          let completion_id = self # next_completion_id in
+          (*
+              (for now) gobjfs does not support arbitrary partial reads
+              but offset and len need to be multiples of 4096
+
+              |---- left ------- | ----- len ----- |----- right -----|
+             off'               off             off+len          off'+len'
+
+           *)
+
+          let left = off mod 4096 in
+          let off' =
+            if left = 0
+            then off
+            else off - left
+          in
+          let remainder = (left + len) mod 4096 in
+          let len' =
+            if remainder = 0
+            then left + len
+            else left + len + (4096 - remainder)
+          in
+          let bytes = GMemPool.alloc len' in
+          let fragment = Fragment.make completion_id off' len' bytes in
+          (completion_id, off', off, len, len', fragment)
+        ) slices
+    in
+    let fragments =
+      List.map
+        (fun (completion_id, off', off, len, len', fragment) -> fragment)
+        corrections
+    in
+    let batch = Batch.make fragments in
+
+    Lwt.finalize
+      (fun () ->
+        Prelude.with_timing_lwt
+          (fun () ->
+            IOExecFile.file_read handle batch _event_channel >>= fun () ->
+            Lwt_list.map_p
+              (fun correction ->
+                let completion_id, _, _, _,_, _ = correction in
+                Prelude.with_timing_lwt (fun () -> self # _wait_for_completion completion_id)
+                >>= fun (took,ec) ->
+                let () = AsdStatistics.new_delta _statistics _READ_WAIT took in
+                Lwt.return (correction, ec)
+              )
+              corrections
+          )
+        >>= fun (took,results) ->
+
+        let () = AsdStatistics.new_delta _statistics _READ_BATCH took in
+
+        _process_results
+          (fun (_,ec) -> ec)
+          "IOExecFile.file_read"
+          (fun _ -> fn)
+          results
+        >>= fun () ->
+        Prelude.with_timing_lwt
+          (fun () -> Lwt_list.iter_s _f results)
+        >>= fun (took,()) ->
+        let () = AsdStatistics.new_delta _statistics _READ_BATCH_CB took in
+        Lwt.return_unit
+      )
+      (fun () ->
+        List.iter Fragment.free_bytes fragments;
+        Prelude.with_timing_lwt
+          (fun () ->IOExecFile.file_close handle)
+        >>= fun (took,()) ->
+        AsdStatistics.new_delta _statistics _FILE_CLOSE took;
+        Lwt.return_unit
+      )
+
+  method push_blob_data fnr len slices f =
```

Comment:
I considered get_blob_data first, but then the 'f' feels out of place. Anyway, I'll change.